pkg/apis/kubeone/v1beta1/types.go

```diff
@@ -0,0 +1,331 @@
+	// DynamicWorkers descirbes the worker nodes that are managed by
+	// Kubermatic machine-controller/Cluster-API
+	DynamicWorkers []DynamicWorkerConfig `json:"dynamicWorkers,omitempty"`
+	// MachineController configures the Kubermatic machine-controller component
+	MachineController *MachineControllerConfig `json:"machineController,omitempty"`
+	// Features enables and configures additional cluster features
+	Features Features `json:"features,omitempty"`
+	// Addons are used to deploy additional manifests
+	Addons *Addons `json:"addons,omitempty"`
+	// SystemPackages configure kubeone behaviour regarding OS packages
+	SystemPackages *SystemPackages `json:"systemPackages,omitempty"`
+}
+
+// OperatingSystemName defines the operating system used on instances
+type OperatingSystemName string
+
+var (
+	OperatingSystemNameUbuntu  OperatingSystemName = "ubuntu"
+	OperatingSystemNameCentOS  OperatingSystemName = "centos"
+	OperatingSystemNameCoreOS  OperatingSystemName = "coreos"
+	OperatingSystemNameFlatcar OperatingSystemName = "flatcar"
+	OperatingSystemNameUnknown OperatingSystemName = ""
+)
+
+// HostConfig describes a single control plane node.
+type HostConfig struct {
+	ID                int    `json:"-"`
+	PublicAddress     string `json:"publicAddress"`
+	PrivateAddress    string `json:"privateAddress"`
+	SSHPort           int    `json:"sshPort"`
+	SSHUsername       string `json:"sshUsername"`
+	SSHPrivateKeyFile string `json:"sshPrivateKeyFile"`
+	SSHAgentSocket    string `json:"sshAgentSocket"`
+	Bastion           string `json:"bastion"`
+	BastionPort       int    `json:"bastionPort"`
+	BastionUser       string `json:"bastionUser"`
+	Hostname          string `json:"hostname"`
+	IsLeader          bool   `json:"isLeader"`
+
+	// If not provided (i.e. nil) defaults to TaintEffectNoSchedule, with key
+	// node-role.kubernetes.io/master for control plane nodes.
+	//
+	// Explicitly empty (i.e. []corev1.Taint{}) means no taints will be applied
+	// (this is default for worker nodes).
+	Taints []corev1.Taint `json:"taints,omitempty"`
+
+	// Information populated at the runtime
+	OperatingSystem OperatingSystemName `json:"-"`
+}
+
+// ControlPlaneConfig defines control plane nodes
+type ControlPlaneConfig struct {
+	Hosts []HostConfig `json:"hosts"`
+}
+
+// StaticWorkersConfig defines static worker nodes provisioned by KubeOne and kubeadm
+type StaticWorkersConfig struct {
+	Hosts []HostConfig `json:"hosts"`
+}
+
+// APIEndpoint is the endpoint used to communicate with the Kubernetes API
+type APIEndpoint struct {
+	// Host is the hostname on which API is running
+	Host string `json:"host"`
+
+	// Port is the port used to reach to the API
+	Port int `json:"port"`
+}
+
+// CloudProviderSpec describes the cloud provider that is running the machines.
+// Only one cloud provider must be defined at the single time.
+type CloudProviderSpec struct {
+	External     bool              `json:"external"`
```

Comment:
Hm. This probably also should be a struct.

Reply:
What we would gain from that? And how it would look like actually? From the API usability side, I think having it as a bool would be the best approach. I'd just say that we should make it a named-bool instead.

I've been thinking about putting it into the provider-specific structs that we added. But I have realized that you can choose to use external for any provider, even for AWS, Azure... A top-level field in CloudProviderSpec makes more sense because of that.

Maybe it could be useful to have a struct such as:
```go
type ExternalCloudProviderConfig struct {
    DeployCCM bool `json:"deployCCM"`
}
```
That would allow some additional control for providers with the external CCM templates built-in.

Reply:
A combination of provider=none and external=false gives us what ExternalCloudProviderConfig.DeployCCM can give (and they seems like mutually exclusive), so I'd be against `ExternalCloudProviderConfig`.

Reply:
OK, let it stay a bool